File: src/test/java/kitchenpos/products/tobe/application/ProductServiceTest.java
package kitchenpos.products.tobe.application;

import static java.math.BigDecimal.valueOf;
import kitchenpos.products.tobe.infra.FakeProfanitiesClient;
import kitchenpos.menus.application.InMemoryMenuRepository;
import kitchenpos.menus.domain.MenuRepository;
import kitchenpos.products.tobe.domain.exception.InvalidProductException;
import kitchenpos.products.tobe.domain.vo.Profanities;
import kitchenpos.products.tobe.ui.dto.ChangeProductRequest;
import kitchenpos.products.tobe.ui.dto.ChangeProductResponse;
import kitchenpos.products.tobe.ui.dto.CreateProductRequest;
import kitchenpos.products.tobe.ui.dto.CreateProductResponse;
import kitchenpos.products.tobe.domain.Product;
import kitchenpos.products.tobe.domain.vo.ProductName;
import kitchenpos.products.tobe.domain.vo.ProductPrice;
import kitchenpos.products.tobe.domain.ProductRepository;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;


class ProductServiceTest {

    private ProductService productService;
    private ProductRepository productRepository;
    private MenuRepository menuRepository;
    private Profanities profanities;

    @BeforeEach
    void setUp() {
        productRepository = new InMemoryProductRepository();
        menuRepository = new InMemoryMenuRepository();
        profanities = new FakeProfanitiesClient();
        productService = new ProductService(productRepository, menuRepository, profanities);
    }

    @Test
    void 상품을_등록할_수_있다() {
        // given
        ProductName name = new ProductName("후라이드치킨");
        ProductPrice price = new ProductPrice(valueOf(16000));
        CreateProductRequest request = new CreateProductRequest(name.getName(), price.getPrice());

        // when
        CreateProductResponse savedProduct = productService.create(request);

        // then
        assertThat(savedProduct).isNotNull();
        assertThat(savedProduct.id()).isNotNull();
        assertThat(savedProduct.name()).isEqualTo("후라이드치킨");
        assertThat(savedProduct.price()).isEqualTo(valueOf(16000));
    }
    @Test
    void 상품의_가격을_변경할_수_있다() {
        // given
        Product product = new Product("후라이드치킨", valueOf(16000));
        productRepository.save(product);

        UUID productId = product.getId();
        BigDecimal 변경할_가격 = valueOf(17000);
        ChangeProductRequest request = new ChangeProductRequest(변경할_가격);

        // when
        ChangeProductResponse response = productService.changePrice(productId, request);

        // then
        Product updatedProduct = productRepository.findById(productId).orElseThrow();

        assertThat(updatedProduct.getPrice().compareTo(변경할_가격)).isEqualTo(0);
        assertThat(response.productId()).isEqualTo(productId);
        assertThat(response.price()).isEqualTo(변경할_가격);
    }

    @Test
    void 존재하지_않은_상품_ID로_가격을_변경할_수_없다() {
        // given
        UUID nonExistentProductUd = UUID.randomUUID();
        BigDecimal 변경할_가격 = valueOf(20000);
        ChangeProductRequest request = new ChangeProductRequest(변경할_가격);

        // when & then
        assertThatThrownBy(() -> productService.changePrice(nonExistentProductUd, request))
                .isInstanceOf(InvalidProductException.class);
    }

    @Test
    void 등록된_상품을_전체_조회할_수_있다() {
        // given
        ProductName name = new ProductName("후라이드치킨");
        ProductPrice price = new ProductPrice(valueOf(16000));
        CreateProductRequest request = new CreateProductRequest(name.getName(), price.getPrice());

        // when
        productService.create(request);

        // then
        List<Product> products = productRepository.findAll();
        assertNotNull(products);
        assertThat(products).hasSize(1);
    }
}

File: src/test/java/kitchenpos/products/tobe/application/ProductServiceTest.kt
package kitchenpos.products.tobe.application

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.justRun
import io.mockk.mockk
import io.mockk.verify
import org.springframework.context.ApplicationEventPublisher
import java.math.BigDecimal

val Int.won: BigDecimal get() = BigDecimal(this)

class ProductServiceTest : BehaviorSpec({
    val productRepository = InMemoryProductRepository()
    val eventPublisher = mockk<ApplicationEventPublisher>()
    val productService =
        ProductService(
            productRepository = productRepository,
            productNameValidator = { listOf("유효하지 않은 이름").contains(it).not() },
            eventPublisher = eventPublisher,
        )

    val validName = "유효한 이름"
    val invalidName = "유효하지 않은 이름"
    val validPrice = 1000.won
    val invalidPrice = (-1000).won

    justRun { eventPublisher.publishEvent(any()) }

    Given("상품을 생성할 떄") {
        When("상품 이름과 가격이 유효하면") {
            val newProduct = productService.create(validName, validPrice)

            Then("상품이 성공적으로 생성된다.") {
                newProduct.name shouldBe validName
                newProduct.price shouldBe validPrice
            }
        }

        When("상품 가격이 올바르지 않으면") {
            Then("예외가 발생한다.") {
                shouldThrow<IllegalArgumentException> {
                    productService.create(validName, invalidPrice)
                }
            }
        }

        When("상품 이름이 유효하지 않으면") {
            Then("예외가 발생한다.") {
                shouldThrow<IllegalArgumentException> {
                    productService.create(invalidName, validPrice)
                }
            }
        }
    }

    Given("상품 가격을 변경할 때") {
        When("상품이 존재하는 경우") {
            val existingProduct = productService.create(validName, validPrice)
            val updateProduct = productService.changePrice(existingProduct.id, 2000.won)

            Then("상품 가격이 업데이트된다.") {
                updateProduct.price shouldBe 2000.won
            }

            Then("이벤트가 발행된다.") {
                verify { eventPublisher.publishEvent(any()) }
            }
        }
    }
})
